azure-iot-device/azure/iot/device/common/pipeline/pipeline_thread.py

```diff
@@ -0,0 +1,199 @@
+
+                # TODO: add a timeout here and throw exception on failure
+                future = _get_named_executor(thread_name).submit(thread_proc)
+                if block:
+                    return future.result()
+                else:
+                    return future
+            else:
+                logger.debug("Already in {} thread for {}".format(thread_name, function_name))
+                return func(*args, **kwargs)
+
+        # Silly hack:  On 2.7, we can't use @functools.wraps on callables don't have a __name__ attribute
+        # attribute(like MagicMock object), so we only do it when we have a name.  functools.update_wrapper
+        # below is the same as using the @functools.wraps(func) decorator on the wrapper function above.
+        if getattr(func, "__name__", None):
+            return functools.update_wrapper(wrapped=func, wrapper=wrapper)
+        else:
+            return wrapper
+
+    if _func is None:
+        return decorator
+    else:
+        return decorator(_func)
+
+
+def invoke_on_pipeline_thread(_func=None):
+    """
+    Run the decorated function on the pipeline thread.
+    """
+    return _invoke_on_executor_thread(thread_name="pipeline", _func=_func)
+
+
+def invoke_on_pipeline_thread_nowait(_func=None):
+    """
+    Run the decorated function on the pipeline thread, but don't wait for it to complete
+    """
+    return _invoke_on_executor_thread(thread_name="pipeline", block=False, _func=_func)
+
+
+def invoke_on_callback_thread_nowait(_func=None):
+    """
+    Run the decorated function on the callback thread, but don't wait for it to complete
+    """
+    return _invoke_on_executor_thread(thread_name="callback", block=False, _func=_func)
+
+
+def _assert_executor_thread(thread_name, _func=None):
+    """
+    Decorator which asserts that the given function only gets called inside the given
+    thread.
+    """
+
+    def decorator(func):
+        function_name = getattr(func, "__name__", str(func))
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if not getattr(_get_thread_local_storage(), "in_{}_thread".format(thread_name), False):
+                assert False, """Function {function_name} is not running inside {thread_name} thread.
+                    It should be. You should use invoke_on_{function_name}_thread(_nowait) to enter the
+                    {thread_name} thread before calling this function.  If you're hitting this from
+                    inside a test function, you may need to add the fake_pipeline_thread fixture to
+                    your test.  (grep for apply_fake_pipeline_thread).""".format(
+                    function_name=function_name, thread_name=thread_name
+                )
+
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    if _func is None:
+        return decorator
+    else:
+        return decorator(_func)
+
+
+def runs_on_pipeline_thread(_func=None):
```

Comment:
>runs_on_pipeline_thread [](start = 4, length = 23)

so this in only for assertion ? in that case could we add an `ensure` or `assert`  before ?

Reply:
`ensure` wouldn't be appropriate because that implies that it might change something.  `ensure_running_on_pipeline_thread` would imply a function that checks the thread and change it if necessary.
Either way, I'm going to leave it like this for now.

---
In reply to: [301844700](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r301844700) [](ancestors = 301844700)